### PR TITLE
Gradient accumulation fix in cross entropy loss

### LINKED ISSUE #20350: Gradient accumulation calcluation may be incorrect
### Bug description

See https://unsloth.ai/blog/gradient for in depth explanation, and [this PR](https://github.com/huggingface/transformers/pull/34191/files) for how huggingface fixed it.

I verified that I see worse performance with gradient accumulation than multiple devices, so I suspect this bug also applies to Lightning

### What version are you seeing the problem on?

v2.4

### How to reproduce the bug

_No response_

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.4.0):
#- PyTorch Version (e.g., 2.4):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by Sohaib-Ahmed21**
## What does this PR do?

This PR resolves the issue where loss normalization during gradient accumulation is incorrect in case of cross entropy loss.

**Key Changes**:

- Implemented a **peekable iterator** that inspects all microbatches in the global batch before the first forward pass in accumulation window to determine the total number of valid tokens.

- Added support for a `labels` key in each batch, similar to the `labels` field used in Hugging Face Transformers.

- Each microbatch’s loss is now divided by the total valid-token count of the global batch, ensuring correct scaling during gradient accumulation.

- Documentation updates will follow to guide users on required batch structure and loss settings.

Fixes #20350 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- yes
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
yes will do so.
- Did you write any **new necessary tests**? (not for typos and docs)
I think so not needed yet.
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21386.org.readthedocs.build/en/21386/

### CODE REVIEW

File: src/lightning/pytorch/loops/training_epoch_loop.py [hunk 1]
```diff
@@ -346,6 +355,38 @@ def advance(self, data_fetcher: _DataFetcher) -> None:
...
+                    peek_batches = list(islice(self._peekable_iter, trainer.accumulate_grad_batches))
+                    for batch in peek_batches:
+                        # unwrap Lightning's list/tuple wrapper
+                        if isinstance(batch, (list, tuple)):
+                            batch = batch[0]
+                        # require batch to be instance of dict and has labels, otherwise break
+                        if not isinstance(batch, dict):
+                            all_formatted_batches = False
+                            break
+                        labels = batch.get("labels")
```
Comment (SkafteNicki) [inline src/lightning/pytorch/loops/training_epoch_loop.py]: so the core assumption to get this working is that the user have formatted their batches such that each batch has a labels tensor?

File: src/lightning/pytorch/loops/training_epoch_loop.py [hunk 2]
```diff
@@ -346,6 +355,38 @@ def advance(self, data_fetcher: _DataFetcher) -> None:
...
+                        # break if labels missing or None
+                        if labels is None:
+                            all_formatted_batches = False
+                            break
+                        # safe to process
+                        labels = torch.as_tensor(labels)
+                        total_valid_tokens += int((labels != -100).sum().item())
+                    self._num_global_valid_tokens = total_valid_tokens if all_formatted_batches else None
```
Comment (SkafteNicki) [inline src/lightning/pytorch/loops/training_epoch_loop.py]: i do wonder if this is such a special case that it is better for the user to provide the information compared to lightning trying to calculate it. For example, what if the masking token is not `-100`  but `-1` for example?

File: src/lightning/pytorch/loops/training_epoch_loop.py [hunk 3]
```diff
@@ -278,6 +281,12 @@ def advance(self, data_fetcher: _DataFetcher) -> None:
             StopIteration: When the epoch is canceled by the user returning -1
 
         """
+        # create a peekable iterator to look ahead without consuming the original data_fetcher
+        iterator = data_fetcher.iterator
+        assert iterator is not None
+        it1, self._peekable_iter = itertools.tee(iterator)
```
Comment (SkafteNicki) [inline src/lightning/pytorch/loops/training_epoch_loop.py]: this definitely seems like the smart way to implement this kind of solution

File: src/lightning/pytorch/loops/training_epoch_loop.py (hunk 1 above)
Comment (Sohaib-Ahmed21) [reply src/lightning/pytorch/loops/training_epoch_loop.py]: Yes, this will be listed in relevant docs (once the approach is confirmed) and its the same as `labels` key in [transformers](https://github.com/huggingface/transformers).

File: src/lightning/pytorch/loops/training_epoch_loop.py (hunk 2 above)
Comment (Sohaib-Ahmed21) [reply src/lightning/pytorch/loops/training_epoch_loop.py]: For this, I thought that if a user is passing `labels` for this special case, then we also list in the same relevant docs that masking token should be `-100` for this case so users don't miss it. Thoughts please?

**Comment by Sohaib-Ahmed21** [comment]
@SkafteNicki @justusschock kindly review this PR when you have some time so I can proceed with it, thanks!

**Comment by Sohaib-Ahmed21** [comment]
@SkafteNicki @justusschock This PR is ready for review. Kindly review it, thanks!

**Comment by Sohaib-Ahmed21** [comment]
@SkafteNicki @lantiga @justusschock @deependujha all checks are passing now and the PR is ready for review. Kindly review it, thanks!

**Comment by deependujha** [comment]
Hi @Sohaib-Ahmed21, thanks for digging into this.

I think this is a case where Lightning should avoid encoding domain assumptions and instead expose a more general mechanism.

Assumptions around `CrossEntropyLoss`, ignore indices like `-100`, and how “valid tokens” are inferred are very NLP- and HF-specific. Lightning, on the other hand, is intentionally domain-agnostic and used across vision, speech, RL, and custom losses. Once Lightning starts inferring `valid_token_count` from labels, it implicitly assumes token-level supervision, a specific loss function, and specific masking semantics. That feels brittle and difficult to generalize.

---

A cleaner approach would be to **`let the user explicitly control normalization via the training_step output`**.

> For example, the user could return a generic `normalize` value alongside the loss, and Lightning would simply divide the loss by this value during gradient accumulation.

```python
def training_step(...):
    return {"loss": loss, "normalize": valid_tokens}
```

In the case of token-level cross entropy, `normalize` would naturally be the number of valid (non-masked) tokens. In other setups, it could represent batch size, number of sequences, number of samples seen, or any other user-defined notion of “effective batch”.

This keeps responsibilities clearly separated:

* the user defines loss semantics and masking
* Lightning handles accumulation and scaling, without needing to know *why* a given normalization is correct

> [!IMPORTANT]
> Documentation and examples can still recommend returning `normalize=valid_token_count` for HF-style CE with `ignore_index=-100`, but the framework itself remains free of hardcoded assumptions. This feels more robust, extensible, and aligned with Lightning’s design philosophy than inferring normalization internally.


What do you think @SkafteNicki @justusschock @bhimrazy @Sohaib-Ahmed21?

**Comment by Sohaib-Ahmed21** [comment]
Thanks for the detailed feedback @deependujha.

The same thoughts were shared by @SkafteNicki [here](https://github.com/Lightning-AI/pytorch-lightning/issues/20350#issuecomment-3384709701). I followed the [HF implementation](https://github.com/huggingface/transformers/pull/34191) but the scope creep in case of **pytorch-lightning** as you mentioned above seems intuitive. 

I think in the light of comments by you and @SkafteNicki, I should revert back my logic to contain something like `num_ignore_val` in `training_step` output and mention in docs accordingly. Thoughts please @deependujha @SkafteNicki, thanks!

**Comment by deependujha** [comment]
why `num_ignore_val`?

**Comment by Sohaib-Ahmed21** [comment]
> why `num_ignore_val`?

Sorry the name is not descriptive, will have some definitive param to denote total count of valid tokens in global batch.

Can you kindly confirm the approach so I revert, thanks!

**Comment by deependujha** [comment]
This approach makes minimal assumptions and addresses the gradient accumulation bug in the cross-entropy loss without requiring significant code restructuring. It would be good to align on this approach.z

thoughts @bhimrazy ?

**Comment by Sohaib-Ahmed21** [comment]
I've reverted the PR to align on the discussed approach. @deependujha @justusschock @SkafteNicki please review it.

**Comment by bhimrazy** [comment]
> This approach makes minimal assumptions and addresses the gradient accumulation bug in the cross-entropy loss without requiring significant code restructuring. It would be good to align on this approach.z
> 
> thoughts @bhimrazy ?

yeah like the idea @deependujha. 

Also would be good to get a lead's review on approach & to catch any edge cases we might be missing.

File: src/lightning/pytorch/loops/optimization/automatic.py [hunk 4]
```diff
@@ -78,6 +78,13 @@ def from_training_step_output(cls, training_step_output: STEP_OUTPUT, normalize:
             )
 
         if closure_loss is not None:
+            # Override normalization if specified (fixes CrossEntropyLoss accumulation)
+            if (
+                normalize > 1
+                and isinstance(training_step_output, Mapping)
+                and training_step_output.get("normalize") is not None
+            ):
+                normalize = training_step_output["normalize"]
```
Comment (bhimrazy) [inline src/lightning/pytorch/loops/optimization/automatic.py:87]: It will probably need some validations here (e.g., checks for 0 and type checks).

cc: @deependujha

**Comment by Sohaib-Ahmed21** [comment]
> Also would be good to get a lead's review on approach & to catch any edge cases we might be missing.

@SkafteNicki @tchaton @Borda  please review this PR, thanks!

**Comment by Sohaib-Ahmed21** [comment]
Someone please review this.

**Comment by Sohaib-Ahmed21** [comment]
Review on this PR is pending for so long, someone please review @SkafteNicki
